icu4c/source/i18n/units_complexconverter.cpp

```diff
@@ -21,32 +21,63 @@
 
 U_NAMESPACE_BEGIN
 namespace units {
+ComplexUnitsConverter::ComplexUnitsConverter(const MeasureUnitImpl &inputUnit,
+                                             const ConversionRates &ratesInfo, UErrorCode &status): units_(inputUnit.extractIndividualUnitsWithIndecies(status)) {
+    
+        if (U_FAILURE(status)) {
+            return;
+        }
+
+            U_ASSERT(units_.length() != 0);
+
+    auto singleUnits = inputUnit.extractIndividualUnits(status);
+    if (U_FAILURE(status)) {
+        return;
+    }
+
+    auto *biggestUnit = singleUnits[0];
+    for (int32_t i = 1; i < singleUnits.length(); i++) {
+        if (UnitConverter::compareTwoUnits(*singleUnits[i], *biggestUnit, ratesInfo, status) > 0 &&
+            U_SUCCESS(status)) {
+            biggestUnit = singleUnits[i];
+        }
+
+        if (U_FAILURE(status)) {
+            return;
+        }
+    }
+
+  this->init(*biggestUnit, ratesInfo, status);
+}
 
 ComplexUnitsConverter::ComplexUnitsConverter(const MeasureUnitImpl &inputUnit,
                                              const MeasureUnitImpl &outputUnits,
                                              const ConversionRates &ratesInfo, UErrorCode &status)
-    : units_(outputUnits.extractIndividualUnits(status)) {
+    : units_(outputUnits.extractIndividualUnitsWithIndecies(status)) {
     if (U_FAILURE(status)) {
         return;
     }
 
     U_ASSERT(units_.length() != 0);
 
-    // Save the desired order of output units before we sort units_
-    for (int32_t i = 0; i < units_.length(); i++) {
-        outputUnits_.emplaceBackAndCheckErrorCode(status, units_[i]->copy(status).build(status));
-    }
+    this->init(inputUnit, ratesInfo, status);
+}
```

Comment:
This constructor is now broken or misleading: it is simply ignoring the outputUnits parameter? Rather delete it if it isn't useful anymore.

Reply:
why it is broken or misleading??? we need both of the constructors!

one for with only input units and the other for input and output units

Reply:
My mistake: I saw outputUnits wasn't being passed to "init" and thus thought it was ignored - however it's handled via units_.

So that's maybe a way of making the code clearer too: passing outputUnits to init as a parameter instead of via the units_ member. Considering init() is in any case modifying units_, thus taking responsibility for it, this code division makes that feel like a side-effect rather than a clearly declared responsibility. As a parameter, it's more explicitly letting init() take care of it?

Reply:
I'm going to move this "side effect" to be a aprameter of init.

Reply:
I went in a different direction with a change in the three-parameter constructor: reusing units_ in determining a parameter to pass to init(), so I can't assign units_ in init(). Leaving this as "a side-effect", we're not writing in a functional style anyway. :)